frontend/src/metabase/querying/components/FilterModal/FilterModal.tsx

```diff
@@ -53,22 +54,25 @@ export const useFilterContent = (
   const handleChange = (newQuery: Lib.Query) => {
     setQuery(newQuery);
     setIsChanged(true);
+    // for handleSubmit to see the latest query if it is called in the same tick
+    queryRef.current = newQuery;
```

Comment:
I removed this line and used `useLatest` instead of `useRef` - everything seems to work.

`useLatest` is more future-proof than manually manipulating `queryRef.current` on every change, so I suggest to use that instead.

Reply:
Did it pass the e2e test? I also used it initially and in some cases it didn't work. At least that's what I tried:
```
const [query, setQuery] = ...;
const queryRef = useLatest(query);

const handleChange = (newQuery) => {
  setQuery(newQuery);
};

const handleSubmit = () => {
  onSubmit(queryRef.current);
};
```

Reply:
I only checked unit tests modified in this PR.

Reply:
The issue cannot be reproduced in unit tests

Reply:
That's actually very bad because it means we need to rely on e2e more